plugins/catalog-backend/src/ingestion/processors/microsoftGraph/client.ts

```diff
@@ -0,0 +1,232 @@
+        authority: `${config.authority}/${config.tenantId}`,
+      },
+    };
+    const pca = new msal.ConfidentialClientApplication(clientConfig);
+    return new MicrosoftGraphClient(config.target, pca);
+  }
+
+  constructor(
+    private readonly baseUrl: string,
+    private readonly pca: msal.ConfidentialClientApplication,
+  ) {}
+
+  async api(path: string, query?: ODataQuery): Promise<Response> {
+    const queryString = qs.stringify(
+      {
+        $filter: query?.filter,
+        $select: query?.select?.join(','),
+        $expand: query?.expand?.join(','),
+      },
+      {
+        addQueryPrefix: true,
+        // Microsoft Graph doesn't like an encoded query string
+        encode: false,
+      },
+    );
+
+    return await this.request(`${this.baseUrl}/${path}${queryString}`);
+  }
+
+  async request(url: string): Promise<Response> {
+    // Make sure that we always have a valid access token (might be cached)
+    const token = await this.pca.acquireTokenByClientCredential({
+      scopes: ['https://graph.microsoft.com/.default'],
+    });
+
+    return await fetch(url, {
+      headers: {
+        Authorization: `Bearer ${token.accessToken}`,
+      },
+    });
+  }
+
+  async getUserProfile(userId: string): Promise<MicrosoftGraph.User> {
+    const response = await this.api(`users/${userId}`);
+
+    if (response.status !== 200) {
+      await this.handleError('user profile', response);
+    }
+
+    return await response.json();
+  }
+
+  async getUserPhotoWithSizeLimit(
+    userId: string,
+    maxSize: number,
+  ): Promise<string | undefined> {
+    const response = await this.api(`users/${userId}/photos`);
+
+    if (response.status === 404) {
+      return undefined;
+    } else if (response.status !== 200) {
+      await this.handleError('user photos', response);
+    }
+
+    const result = await response.json();
+    const photos = result.value as MicrosoftGraph.ProfilePhoto[];
+    let selectedPhoto: MicrosoftGraph.ProfilePhoto | undefined = undefined;
+
+    // Find the biggest picture that is small than the max size
+    for (const p of photos) {
+      if (
+        !selectedPhoto ||
+        (p.height! >= selectedPhoto.height! && p.height! <= maxSize)
+      ) {
+        selectedPhoto = p;
+      }
+    }
+
+    if (!selectedPhoto) {
+      return undefined;
+    }
+
+    return await this.getUserPhoto(userId, selectedPhoto.id!);
+  }
+
+  async getUserPhoto(
+    userId: string,
+    sizeId?: string,
+  ): Promise<string | undefined> {
+    const path = sizeId
+      ? `users/${userId}/photos/${sizeId}/$value`
+      : `users/${userId}/photo/$value`;
+    const response = await this.api(path);
+
+    if (response.status === 404) {
+      return undefined;
+    } else if (response.status !== 200) {
+      await this.handleError('photo', response);
+    }
+
+    return `data:image/jpeg;base64,${Buffer.from(
+      await response.arrayBuffer(),
+    ).toString('base64')}`;
```

Comment:
Hm ok, this is cool but will increase the entity data size significantly. Would be nice to have some way to dump these blobs separately (a facility to host things like this)

Reply:
I can't "hot link" the images at the Microsoft Graph API as they need auth so I have to copy them. 

The data size is already a problem. Our org (and not everyone has a photo) has user entities with a size of around 9 mb. Therefore I limited the size to max 120x120px, now it's less than a megabyte. 

A way to store photo in a separate entity would be cool. Or not to include them in the default entity response but have another endpoint to access them would be fine too. We noticed the same thing at our auth endpoint where we are including a big photo.

If it's not part of backstage, we could also copy them to S3 and include S3 urls in the entities. But if we have some kind of blob store in Backstage we could back it with S3 too.

Looking forward for group photos 😆

Reply:
Yeah exactly, a simple blob api for auxiliary data around entities - how about API definitions as well for example?

Another size hog is the membership and hierarchy of users / groups. Transitive membership should probably not be stored (even as relations) since the org root would hold the entire freaking company staff etc. At least we plan to remove ancestors and descendants, and perhaps let a processor fill in the org hierarchy relations as necessary based on whatever appeared in the spec sections.

Reply:
> API definitions as well for example

Interesting point, I agree they can become quite large.